### PR TITLE
Update go-ipld-git to a go-ipld-prime codec

### LINKED ISSUE #45: Create IPLD-Prime codec that can surplant the ipld-format functionality in this repo
Currently, this implements an ipld-format view of blocks containing git data.
We need an [ipld-prime](https://github.com/ipld/go-ipld-prime) view of the same data if we want integrate it into IPFS

### PR DESCRIPTION
**Pull request by willscott**
The existing go-ipld-git codec allows for translation between git file data, and go-ipld-format node representations of the IPLD dag. This work attempts to instead provide a go-ipld-prime representation of that dag.

* [x] Decoder codec (git -> ipld-prime)
* [x] Encoder code (ipld-prime -> git)
* [x] Round-trip tests

fix #45

### CODE REVIEW

File: blob.go
Comment (mvdan) [inline blob.go:46]: unlike a read, a write must process the entirety of the bytes to succeed, so this can be just `w.Write(b)` and you avoid allocating a bytes.Buffer and adding its own overhead.

File: personinfo.go
Comment (mvdan) [inline personinfo.go:11]: debug line? might as well add the input line as %q in the error string

File: personinfo.go
Comment (mvdan) [inline personinfo.go:24]: nit: %q

File: personinfo.go
Comment (mvdan) [inline personinfo.go:44]: nit: %q (here and elsewhere)

File: commit.go
Comment (mvdan) [inline commit.go:80]: if you use a few global constants, then this gets easier, like `line[len(prefixMergetag):]`.

File: commit.go
Comment (mvdan) [inline commit.go:145]: I'd strongly advise you to use something like strings.Builder instead; each `+=` on a string is an allocation and copy of the entire thing. Just twenty of these operations could already add up to showing up on CPU profiles.

File: commit.go
Comment (mvdan) [inline commit.go:153]: I think you could even skip this check entirely. for example, go-codec-dagpb always assigns the input node to a PBNode, and it's fine since it's a no-op when the input node is already a PBNode.

I imagine it's not anything to worry about performance-wise, as I profiled the codec and never even saw it show up.

File: commit.go
Comment (mvdan) [inline commit.go:162]: if you think this buffer could get large, you could also use a bufio.Writer, which will write to the final writer in chunks. But if we think it will never be more than a few kilobytes, it's likely irrelevant.

File: commit.go
Comment (mvdan) [inline commit.go:166]: you can simplify these by using %x instead :)

File: helper.go
Comment (mvdan) [inline helper.go:16]: I assume we want to keep this for backwards compatibility. Do we want to at least deprecate it?

File: ipld_missing.go
Comment (mvdan) [inline ipld_missing.go:18]: Hm, I don't follow how this is useful or why we need it.

File: personinfo.go
Comment (mvdan) [inline personinfo.go:31]: same as before, even if we think names should be short :)

File: personinfo.go
Comment (mvdan) [inline personinfo.go:85]: clever - though you might find that just using a `strings.Builder` directly is easier :)

File: tag.go
Comment (mvdan) [inline tag.go:42]: the comments I made earlier about constants could apply here too

File: tag.go
Comment (mvdan) [inline tag.go:103]: ditto

File: tag.go
Comment (mvdan) [inline tag.go:169]: any particular reason you write the word-length-null directly to w, but the rest via buffer? I think you should do everything the same way, be it direct writes, a bytes.Buffer, or a bufio.Writer. It's also a bit confusing that the writes seem to happen out of order, because you fill the buffer first but flush it later.

(this applies to all other uses of buf.WriteTo too)

File: tree.go
Comment (mvdan) [inline tree.go:16]: I've seen this "read number string until null, strip null, parse number" a bunch of times. I think it's enough lines to split it out, like `readNullTerminatedNumber` :)

File: tree.go
Comment (mvdan) [inline tree.go:27]: I'm not sure I follow why we have a Tree__Repr builder here, and its result is built and assigned into a Tree builder, which is then again built. It seems like the other decode funcs follow the same pattern. Wouldn't a single Build per Decode be enough?

File: tree.go
Comment (mvdan) [inline tree.go:33]: same as before; why not pass the `la.AssembleValue()` assembler to the decoder and let it assign directly into it?

File: tree.go
Comment (mvdan) [inline tree.go:40]: reminder that AssignNode returns an error, even if you think that these should never fail

File: util.go
Comment (mvdan) [inline util.go:17]: Hm, isn't this a bit wasteful if we only end up using part of the result? Is this to strip the multihash prefix? Might be good to document it a bit.

**Review by mvdan — APPROVED** [review]
SGTM, though take it with a grain of salt as this is the first time I look into git object formats. I also assume that the existing test suite is extensive enough to give us confidence of no regressions.

File: commit.go
Comment (willscott) [reply commit.go:145]: this is copied verbatim from https://github.com/ipfs/go-ipld-git/blob/master/git.go#L310

I'll file as an issue to optimize if we end up caring

File: commit.go
Comment (willscott) [reply commit.go:166]: i'll leave as is since copied verbatim

File: ipld_missing.go
Comment (willscott) [reply ipld_missing.go:18]: i think it was for debugging, but doesn't seem harmful.

File: tag.go
Comment (willscott) [reply tag.go:169]: i guess we first build the object, and only then can generate the prefix with the length of it, so we don't how long  of a prefix to allocate at the beginning of the buffer until the end.

Comment (mvdan) [reply tag.go:169]: Oh, right, of course :) That also kinda disproves the point about bufio.Writer, as it doesn't keep track of written bytes.

File: tree.go
Comment (willscott) [reply tree.go:27]: i think this came out of tests:
the default is to use a basic.any as the assembler, and if that was directly used to make a list / map you end up with a top level basic node rather than the schema'd type that we have in this package.
When that's true, the decode path can't fast-path, and has to re assign that basic node into the schema type.

File: util.go
Comment (willscott) [reply util.go:17]: right. this just uses the sha1 part of the multihash without the prefix to provide the git-compatible hash.

**Comment by rvagg** [comment]
@willscott @mvdan this codec changes all of the struct fields to camel-case, `object` -> `Object`, etc. Would you mind if we change it back for consistency with the old codec and the JS one? It's also closer to the native format's idea of naming as well. I can make the change here if you're both OK with that.

**Comment by rvagg** [comment]
Some additional thoughts here where the differences in forms can be seen from a pathing perspective: https://github.com/ipfs/go-ipfs/pull/8287

If you're OK with it, I'd like to try and get the forms here back to the original.

**Comment by mvdan** [comment]
That sounds fine to me - I'm not sure if ipld-prime codegen expects capitalized names, for the sake of exported names in the generated Go code.

**Comment by willscott** [comment]
I also have no problem with changing name case back for consistency

**Comment by rvagg** [comment]
0436bc7c4 now has the field names matching the original (lower-case, and some other minor adjustments). `tagType` is unfortunate since we can't use `type` in the codegen. `Tree` is now a map too so you can look up by directory name as per the original codec, which lets us path the same way back in the sharness test.

I think we need to do something about this strict need to match Go conventions on these names since it's not always going to match what you want your data model form to look like.

**Comment by mvdan** [comment]
Out of curiosity, would it be enough to add "repr" mapping of field names so that the repr layer uses lowercase names?

Aside from that, I wonder if it's a bit weird for the schema to have the field names be lowercase, but the type names be uppercase.

**Comment by rvagg** [comment]
> Out of curiosity, would it be enough to add "repr" mapping of field names so that the repr layer uses lowercase names?

Sadly, no, because it's pathing at the data model layer that we want to care about, not encoding. Having `rename`s is only going to help if we cared about these objects going into a standard codec (dag-cbor, dag-json), otherwise they're kind of useless here.

**Comment by rvagg** [comment]
> I wonder if it's a bit weird for the schema to have the field names be lowercase, but the type names be uppercase.

This seems fine to me, it actually looks pretty reasonable to my eyes (eyes that have been trained on C-family-style and aren't as accustomed to the upper-case first character rule, granted). Have a look at the schemas that Hannah wrote up in https://github.com/ipfs/go-unixfsnode/blob/main/data/gen/main.go (although the schema field names don't match what's being generated, but you can see the same thing going on there).

**Comment by mvdan** [comment]
SGTM, thanks for confirming on both accounts. I've replied to https://github.com/ipld/go-ipld-prime/issues/206 :)

**Comment by rvagg** [comment]
@mvdan @willscott I've brought this up to what I think is a mergable state:

* rebased on master - which brought in a bunch of stuff to get it working with the PL actions workflows
* updated the README
* done some minor jiggering of the types to match the README, and therefore closer to the original, (one rename of "text" to "message" and a shuffle of some of the struct fields, which mostly shouldn't have any impact on anything).
* brought back the JSON tests using dag-json
* updated to latest go-ipld-prime which fixes the typed link bug

and a few other minor things.

The main difference that I'm in two minds about is date handling - we're pulling out `date` and `timezone` and keeping them faithfully in their original format so round-trips work properly. The original codec presented them as a nice stringified `date` but with the original values kept internally.
An approach I've toyed with in the Bitcoin codec (in JS at least and I was thinking of doing the same in Go) was to do both - keep originals and present the nice usable forms. It just raises questions about what happens when/if they get out of sync. Two possible approaches would be: (1) error if they're out of sync when serializing, or (2) ignore the usable forms entirely when serializing.

For now I think it'd be fine to just leave this as is and leave it to the user to reform the stringified version if they want it, we could add back in a utility function to do that if needed, or evolve this code further in the future.

IMO this can be merged now.

File: gen/gen.go
Comment (willscott) [inline gen/gen.go]: why did you change this to a map rather than a stringjoin?

**Review by willscott — COMMENTED** [review]
all of these changes, mostly the lower-casing of fields, looks good to me.

We should get some sort of nod from stewards before merging, i think?

File: gen/gen.go
Comment (rvagg) [reply gen/gen.go]: Because it's not being used as a stringjoin in any way? We have a custom `PersonInfo#GitString()` that's used to encode it in the git format and `parsePersonInfo()` to parse the input string for decoding. Stringjoin here is only going to mean that when it's encoded with dag-cbor or dag-json that it comes out as `<date> <timezone> <email> <name>`, which IMO is lossy and probably not what we're wanting out of this. Unless I'm missing something?

Comment (willscott) [reply gen/gen.go]: `<date> <timezone> <email> <name>` is the git format though. I would hope that wouldn't be lossy.

Comment (rvagg) [reply gen/gen.go]: it's `$name <$email> $maybedate $maybetimezone` - it's lossy because the email has a `<>` around it and we can use the `<` to delineate the name which itself can contain any number of spaces, so simple stringjoin isn't enough

Comment (warpfork) [reply gen/gen.go]: Agree this seems better to do at the codec layer and yield up to data model as a map.

If I was going to transliterate some git data to dag-json, I'd probably want this to become a map in dag-json.

File: README.md
Comment (mvdan) [inline README.md:34]: since you're not keeping dates in a human-friendly single string format, I wonder if you could make this unix timestamp an Int too.

**Review by mvdan — APPROVED** [review]
LGTM

File: README.md
Comment (aschmahmann) [inline README.md:103]: Is this the only other change to the output data (i.e. breaking changes a user might need to know about)?

From what I can tell from the PR the reason for this change has little to do with UX and is due to some tooling problems on our part. We can potentially live with this, but it'd be helpful to know how painful the fix is.

**Review by aschmahmann — COMMENTED** [review]
Mostly looks good to me, left a few comments/questions. I'm sure you folks have much greater context on the go-ipld-prime usage than I do.

It looks like there are some breaking changes to the data output by this codec. We should be dropping the list of breaking changes into the release tag for when we release this so I think it'd be great if we could writeup the list of breaking changes to the codec (at the spec level) for anyone who happens to be affected. For the purposes of review having an explanation of why a breaking change was made would also be appreciated.

I don't know if we have guidelines on how codec names should match to the internal names of the format we're representing with IPLD, but I'd suspect in general we'd try to match the native names.

File: README.md
Comment (mvdan) [reply README.md:103]: the tagType unfortunate side effect is mentioned here: https://github.com/ipfs/go-ipld-git/pull/46#issuecomment-884854758

perhaps @warpfork has thoughts. it seems to me like it's worth trying to undo this breaking change by improving our codegen before we merge this. I don't have the bandwidth for it today, but I could take a look over the weekend.

File: tree.go
Comment (warpfork) [reply tree.go:19]: I think I agree with mvdan here -- it's a little odd to have a concrete reference like this in a method that is already accepting a NodeAssembler as a parameter.  When there is a NodeAssembler as a parameter, I would usually expect it to just be used directly.  (And maybe documentation can state something about "it is recommended to use a Tree__ReprAssembler here for best results".)

This doesn't seem dangerously incorrect as written, but does seem a little baroque.

**Review by warpfork — APPROVED** [review]
Big picture, LGTM.

+1 big happy on the fix to capitizalization (field names being downcase).  That makes more sense to me.

File: README.md
Comment (warpfork) [reply README.md:103]: Fairly sure this should be doable -> https://github.com/ipld/go-ipld-prime/issues/206#issuecomment-889903156

**Comment by willscott** [comment]
Making sure we've got all the loose ends covered here:

* i looked at the at date handling code again. I think i prefer the current splitting out of timezone in the data model, and i believe it actually is backwards compatible - previous codec encodings will have the full date (with timezone) in their date field, and the prime codec will take that encoding and produce the correct git representation of that data. 
* We should see if we can clean up the the `tagType` field nomenclature per https://github.com/ipld/go-ipld-prime/issues/206#issuecomment-889903156

**Comment by aschmahmann** [comment]
Yes, and a review for other breaking changes. `mergeTag` seems like a breaking change that seems like something we can just restore. I'm not sure if there are others at the moment.

Posting what the release notes for this release would look like would also be great since as soon as this is merged we're going to cut a release anyway and that release is going to need release notes 😄.

**Comment by rvagg** [comment]
Have fixed up `tagType` to `type` thanks to Eric's hint, and `mergeTag` is now `mergetag`, plus a few more readability and clarity tweaks.

Original code has a `MergeTag` as well as a `Tag` with the main difference that one has a `text` field and the other a `message` field but otherwise they are the same. This one just has `Tag` now with `message` which I think is more in line with the git model. That'll be a breaking change from a DAG traversal perspective so we should note it in the release notes.

Will put together some basic release notes next week.

**Comment by rvagg** [comment]
Release notes:

---

BREAKING CHANGES

* go-ipld-git is now a [go-ipld-prime](https://github.com/ipld/go-ipld-prime) IPLD codec. Use `Decode(na ipld.NodeAssembler, r io.Reader) error` and `Encode(n ipld.Node, w io.Writer) error` for direct use if required.
* There is now only a `Tag` type, `MergeTag` has been removed which had a `text` property. Use `Tag`'s `message` property instead to retrieve the tag message from a commit's `mergetag`. i.e. `<commit>/mergetag/message` instead of `<commit>/mergetag/text`.
* `PersonInfo` no longer exposes the human-readable RFC3339 format `date` field as a DAG node. The `date` and `timezone` fields are kept as their original string forms (to enable precise round-trips) as they exist in encoded Git data. e.g. `<commit>/author/date` now returns seconds in string form rather than an RFC3339 date string. Use this value and `<commit>/author/timezone` to reconstruct the original if needed.

**Comment by warpfork** [comment]
Might I suggest we drop that change info into a `CHANGELOG.md` file right in the diff?  PRs disappear from memory quickly and aren't easily indexed over.

(I'm pretty happy with the loose format in [go-ipld-prime/CHANGELOG.md#released-changes](https://github.com/ipld/go-ipld-prime/blob/master/CHANGELOG.md#released-changes), if having an outline to riff from greases the wheels. :))

File: README.md
Comment (warpfork) [reply README.md:103]: I think the schema and code here is now fixed, but this documentation diff needs to be reverted now to match again :)

**Comment by warpfork** [comment]
(Informational note, no action required: I found myself nerdsniped by `mergetag`.  Yes, that's a thing.  Yes, keeping it alllowercase is correct.  Docs here: http://github.com/git/git/blob/master/Documentation/technical/signature-format.txt .)

**Comment by rvagg** [comment]
Have added a changelog and fixed the readme bork.

Normally I don't like changelogs because it's too easy for them to get out of sync, but recently I've adopted the release-on-every-merge practice in JS and it includes both tag + github release + changelog updates and all I need to do is make sure my commit messages are informative and standardised, e.g. https://github.com/rvagg/cborg/blob/master/CHANGELOG.md & https://github.com/rvagg/cborg/releases & https://www.npmjs.com/package/cborg all from pressing the merge button. Can recommend, although Go folks have other things to worry about in the release process I suppose.

File: README.md
Comment (rvagg) [reply README.md:65]: (delayed reply to this) the main idea here is to keep the data lossless and not do too many tricks to make it palatable to the user.

Encoded form looks something like:

```
author A U Thor <author@example.com> 1465981137 +0000
```

But there's even some flexibility in what those final two fields can contain and apparently whether we get two, one or zero of them (I'm not sure about that detail though). So we treat them as strings and present them to the user as they are.

The current version of this codec keeps them internally but only presents a nice human-readable form as a DAG node, which isn't going to work when we're treating DAG traversal as traversal over the data model like we're doing with ipld-prime.

We did discuss (in this thread and other places) some options for making the human-readable form _emerge_ out of the data to be available, but @willscott made the case that since this codec is a bit of a reference codec that we'd want to point others to for how to implement one that we'd better keep it simple for now.

So, lossless data, pure data model, no trickery, user gets to interpret the nodes as they want.

**Comment by warpfork** [comment]
Sweet.  I think this is officially ready for the mergeparty (later this week? early next week?) where we land a bunch of these similar PRs in other repos at once.

File: init.go
Comment (aschmahmann) [inline init.go:20]: Is this necessary for functioning, or just a convenience function for updating the global registry via module imports?

@warpfork I recall your aversion to global registries. Does encouraging codec creators to do this, but codec users to use custom LinkSystems when needed seem like a reasonable compromise?

Comment (willscott) [reply init.go:20]: i believe this was the compromise we ended up at, yes

Comment (warpfork) [reply init.go:20]: Yep, that's exactly it.

Global registries have downsides, but using package-init-time's natural synchronization is just too valuable.  And having ways to opt-out of using the global registry, if some really needs to, seems to mitigate the biggest downsides handily.  So: :+1:

**Comment by rvagg** [comment]
I have created a draft release & tag for this, `v0.1.0`: https://github.com/ipfs/go-ipld-git/releases

I'm working on some assumptions here about Go (and PL) release practices that may not be quite correct. So someone correct me in the next day or two before I press merge and create the release & tag. I would prefer to use semver properly and do a v1.0.0 but I don't think we're in the habit of doing that with Go so I'm assuming v0.1.0 is appropriate, a tag is sufficient and a GitHub release is a bonus.

**Pull request merged**
